pkg/apps/strategy/support/hooks.go

```diff
@@ -0,0 +1,214 @@
+	DeployerPodTimeoutError = fmt.Errorf("timeout while waiting for the deployer pod to start")
+)
+
+func NewLifecycleHookController(kubeclientset kubernetes.Interface, imageClient imageclientv1.Interface, out io.Writer) *LifecycleHookController {
+	c := &LifecycleHookController{
+		kubeClient:  kubeclientset,
+		imageClient: imageClient,
+		out:         out,
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	if glog.V(5) {
+		eventBroadcaster.StartLogging(glog.Infof)
+	}
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+
+	c.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "deployer"})
+
+	return c
+}
+
+func (c *LifecycleHookController) RunHook(hook *appsapi.LifecycleHook, rc *corev1.ReplicationController, hookType string, timeout time.Duration) error {
+	if len(hook.TagImages) > 0 {
+		h := &lifecycleTagHook{
+			imageClient: c.imageClient,
+			hookType:    hookType,
+			out:         c.out,
+		}
+		return h.run(hook, rc)
+	}
+
+	h := &lifecyclePodHook{
+		rcName:            rc.Name,
+		rcNamespace:       rc.Namespace,
+		podClient:         c.kubeClient.CoreV1().Pods(rc.Namespace),
+		hookPodName:       apihelpers.GetPodName(rc.Name, hookType),
+		out:               c.out,
+		hookType:          hookType,
+		deployerStarted:   make(chan struct{}),
+		hookPodStarted:    make(chan struct{}),
+		hookCompleteChan:  make(chan struct{}),
+		workqueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pods"),
+		outputWriterMutex: sync.Mutex{},
+	}
+
+	// Start new filtered informer for all deployer managed pods
+	informer := coreinformers.NewFilteredPodInformer(c.kubeClient, rc.Namespace, 60*time.Second, cache.Indexers{}, func(options *metav1.ListOptions) {
+		options.LabelSelector = labels.SelectorFromSet(map[string]string{appsapi.DeployerPodForDeploymentLabel: rc.Name}).String()
+	})
+
+	h.podLister = v1.NewPodLister(informer.GetIndexer())
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    h.enqueue,
+		UpdateFunc: func(old, new interface{}) { h.enqueue(new) },
+	})
+
+	stopChan := make(chan struct{})
+	defer close(stopChan)
+
+	go informer.Run(stopChan)
+	defer h.workqueue.ShutDown()
+
+	// TODO: Should probably add timeout to prevent this hanging when the caches cannot be synced
+	if !cache.WaitForCacheSync(stopChan, informer.HasSynced) {
+		return h.error("wait for cache sync failed")
+	}
+
+	// Run just one worker as we want to avoid parallel processing
+	go wait.Until(h.runWorker, time.Second, stopChan)
+
+	// Block until we know the the deployer pod has started
+	deployerStartTime, err := h.waitForDeployerPodStarted(timeout, h.deployerStarted)
+	if err != nil {
+		return h.error(err.Error())
+	}
+
+	// We need deployerStartTime so we can set the ActiveDeadlineSeconds for the hook pod which should not account
```

Comment:
I think deployerStartTime is when it goes to Pending so this will account for it anyways